rfcs/20220404-troubleshooting-helpers/README.md

```diff
@@ -0,0 +1,274 @@
+Currently, we already have such debugging facilities all over the
+place, and they are mostly controlled with dedicated macros, that are
+implementation specific (to name a few, we have `GEN_CONV_DEBUG`,
+`TR_DEBUG`, `DEBUGPRINT`).
+
+Here we propose to unify those macros. The main benefit is, for
+developers, that the mechanism to print and debug would be the same
+across implementations.
+
+This macro should be independent of `NDEBUG` as we might need to use
+this logging information on the release build. We propose to call this
+unified macro `ONEDNN_DEVEL_MODE`, and it should have a matching cmake
+option.
+
+We recommend that all messages related to internal logic go under
+`ONEDNN_DEVEL_MODE`. Only messages related to user visible APIs which
+are actionable by user should live outside `ONEDNN_DEVEL_MODE`.
+
+## 2.3 Verbose levels vs verbose flags
+
+Currently, we have some documented information messages we provide
+(primitive creation/execution time), and some undocumented
+(e.g. runtimes info and implementation details).  For the undocumented
+ones, we don't have alignment on which level they should print (some
+print at level 2 and above, other print at level 5 and above).
+
+The current verbose levels were designed with the assumption that each
+level of information was inclusive of the previous ones. If we want to
+increase the number of categories of information to print
+(e.g. dispatch info, perf, error info, debug info, ...), it will
+quickly bloat the verbose output, which can become an issue when using
+verbose for long running workloads.  More importantly, verbose
+messages might impact runtime (e.g. getting perf on GPU), so getting
+some information sometimes impacts runtime.
+
+We propose here to use flags in order to limit verbose bloat, and
+allow printing messages without impacting runtime when possible. This
+could take two forms:
+- a list of message types, like `ERROR,PROFILE,DISPATCH,DEVINFO`...
+- a bit mask encoded in an integer, -1 being all messages are
+  displayed. So we would have for example ERROR=1, INFO=2, DISPATCH=4,
+  ... and the user would have to pass the sum of all diagnostics they
+  want (e.g. for `ERROR` and `DISPATCH`, they would have to pass
+  `ONEDNN_VERBOSE=5`).
+
+We would recommend going with the first option as it is simpler on the
+user side. It would make parsing ONEDNN_VERBOSE more involved, but the
+overhead should be negligible as it happens only once upon printing
+the header of the verbose log.
+
+Regarding the categories of information to print, we would recommend the following:
+- `ERROR`, for all error messages. It can be for runtime / third-party
+  components errors, or any library entry-point that return an error
+  status.
+- `DISPATCH`, for all information related to dispatching
+  information. This is mostly to print why each implementation
+  eligible for dispatch did not dispatch.
+- `PROFILE`, `PROFILE_EXEC` and `PROFILE_CREATE`, for all messages
```

Comment:
QQ: what will `PROFILE` cover? Is it a synonym of `PROFILE_CREATE` + `PROFILE_EXEC`?

Reply:
Yes it would be syntactic sugar for `PROFILE_CREATE,PROFILE_EXEC`.